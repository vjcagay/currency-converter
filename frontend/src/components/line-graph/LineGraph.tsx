import Chart from "chart.js/auto";
import React, { useEffect, useRef } from "react";
import styles from "./styles.module.css";

type Dataset = {
  label?: string;
  data: number[];
  fill?: boolean;
  borderColor?: string;
  backgroundColor?: string;
  tension?: number;
};

type Props = {
  className?: string;
  labels: string[];
  datasets: Dataset[];
};

const LineGraph = (props: Props) => {
  const wrapperRef = useRef<HTMLCanvasElement>();
  const chartRef = useRef<Chart>();

  useEffect(() => {
    chartRef.current = new Chart(wrapperRef.current, {
      type: "line",
      data: {
        labels: [],
        datasets: [],
      },
      options: {
        interaction: {
          intersect: false,
        },
        datasets: {
          line: {
            borderWidth: 1,
            pointStyle: "circle",
            tension: 0.1,
            pointRadius: 0,
            pointHoverRadius: 2,
          },
        },
        scales: {
          x: {
            grid: {
              tickColor: "#666666",
              color: "transparent",
            },
            border: {
              display: true,
              color: "#666666",
            },
            alignToPixels: true,
            ticks: {
              align: "inner",
              autoSkip: true,
              maxRotation: 0,
              autoSkipPadding: 40,
            },
          },
          y: {
            beginAtZero: false,
            position: "right",
            alignToPixels: true,
            border: {
              display: false,
              dash: [4, 4],
              width: 1,
            },
            ticks: {
              align: "end",
            },
          },
        },
        plugins: {
          tooltip: {
            backgroundColor: "#ffffff",
            borderColor: "#eaecf0",
            borderWidth: 1,
            bodyColor: "#0c111d",
            titleColor: "#0c111d",
            titleMarginBottom: 8,
            callbacks: {
              title(ctx) {
                // Exchange rate value
                return (ctx[0].raw as number).toString(10);
              },
              label(ctx) {
                const date = new Date(ctx.label);
                const parts = new Intl.DateTimeFormat("en-US", { dateStyle: "full" }).formatToParts(date);
                return parts.reduce((acc, part) => {
                  if (part.type === "month" || part.type === "day" || part.type === "year") {
                    acc += part.value;
                    acc += part.type === "day" ? "," : ""; // Add a comma after the date.
                    acc += " ";
                  }

                  return acc;
                }, "");
              },
            },
          },
        },
      },
    });
  }, []);

  useEffect(() => {
    if (props.labels && props.datasets) {
      chartRef.current.data.labels = props.labels;
      chartRef.current.data.datasets = props.datasets;
      chartRef.current.update();
    }
  }, [props.labels, props.datasets]);

  return <canvas ref={wrapperRef} className={`${styles.wrapper} ${props.className || ""}`}></canvas>;
};

export default LineGraph;
