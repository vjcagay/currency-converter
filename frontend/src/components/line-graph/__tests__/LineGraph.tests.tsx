import { render } from "@testing-library/react";
import React from "react";
import LineGraph from "../LineGraph";

// Used by Chart.js
global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));

describe("<LineGraph />", () => {
  it("should render correctly", () => {
    const { asFragment } = render(<LineGraph labels={["A", "B", "C"]} datasets={[{ data: [1, 2, 3] }]} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
