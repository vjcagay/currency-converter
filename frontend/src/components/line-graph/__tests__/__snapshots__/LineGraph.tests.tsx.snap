// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LineGraph /> should render correctly 1`] = `
<DocumentFragment>
  <canvas
    class="wrapper "
    height="0"
    style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
    width="0"
  />
</DocumentFragment>
`;
